src/neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -144,26 +137,25 @@ private bool SetGasPerBlock(ApplicationEngine engine, BigInteger gasPerBlock)
             if (gasPerBlock < 0 || gasPerBlock > 10 * GAS.Factor)
                 throw new ArgumentOutOfRangeException(nameof(gasPerBlock));
             if (!CheckCommittee(engine)) return false;
+
             uint index = engine.Snapshot.PersistingBlock.Index + 1;
-            GasRecord gasRecord = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_GasPerBlock)).GetInteroperable<GasRecord>();
-            if (gasRecord[^1].Index == index)
-                gasRecord[^1] = (index, gasPerBlock);
-            else
-                gasRecord.Add((index, gasPerBlock));
+            StorageItem entry = engine.Snapshot.Storages.GetAndChange(CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(index), () => new StorageItem(gasPerBlock));
+            entry.Set(gasPerBlock);
             return true;
         }
 
         [ContractMethod(0_01000000, CallFlags.AllowStates)]
         public BigInteger GetGasPerBlock(StoreView snapshot)
         {
-            var index = snapshot.PersistingBlock.Index;
-            GasRecord gasRecord = snapshot.Storages[CreateStorageKey(Prefix_GasPerBlock)].GetInteroperable<GasRecord>();
-            for (var i = gasRecord.Count - 1; i >= 0; i--)
-            {
-                if (gasRecord[i].Index <= index)
-                    return gasRecord[i].GasPerBlock;
-            }
-            throw new InvalidOperationException();
+            return GetSortedGasRecords(snapshot, snapshot.PersistingBlock.Index + 1).First().gasPerBlock;
+        }
+
+        private IEnumerable<(uint index, BigInteger gasPerBlock)> GetSortedGasRecords(StoreView snapshot, uint to)
+        {
+            return snapshot.Storages.FindRange(
+                CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(to).ToArray(),
+                CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(0u).ToArray(), IO.Caching.SeekDirection.Backward)
```

Comment:
```suggestion
                CreateStorageKey(Prefix_GasPerBlock).ToArray(), IO.Caching.SeekDirection.Backward)
```

Keep `FindRange` to exclude the endkey .